artifact/v3/version/version.cpp

```diff
@@ -0,0 +1,124 @@
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <string>
+#include <vector>
+
+#include <common/common.hpp>
+#include <common/expected.hpp>
+#include <common/error.hpp>
+#include <common/json.hpp>
+
+#include <artifact/v3/version/version.hpp>
+
+namespace artifact {
+namespace v3 {
+namespace version {
+
+namespace io = mender::common::io;
+namespace expected = mender::common::expected;
+namespace error = mender::common::error;
+namespace json = mender::common::json;
+
+
+const ErrorCategoryClass ErrorCategory {};
+
+const char *ErrorCategoryClass::name() const noexcept {
+	return "VersionParserErrorCategory";
+}
+
+string ErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	ParseError:
+		return "Parse error";
+	VersionError:
+		return "Wrong Artifact version";
+	FormatError:
+		return "Wrong Artifact format";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(ErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, ErrorCategory), msg);
+}
+
+
+expected::Expected<Version, error::Error> Parse(io::Reader &reader) {
+	// Read in all the bytes into a tmp buffer
+	std::vector<uint8_t> buf(1024);
+	io::ByteWriter bw {buf};
+
+	int version {-1};
+	string format {"unknown"};
+
+
+	auto err = io::Copy(bw, reader);
+	if (err) {
+		return err;
+	}
+
+	auto expected_json = json::LoadFromString(mender::common::StringFromByteVector(buf));
+
+	if (!expected_json) {
+		return MakeError(ParseError, "Failed to parse the version JSON");
+	}
+
+	const json::Json version_json = expected_json.value();
+
+	if (!version_json.IsObject()) {
+		return MakeError(
+			FormatError,
+			"Failed to unmarshal the version from the version JSON data. The received data does not contain a single top-level object");
+	}
+
+	const json::ExpectedJson version_value = version_json.Get("version");
+	if (!version_value) {
+		return MakeError(FormatError, "No version field found");
+	}
+
+	const json::ExpectedInt value_version = version_value.value().GetInt();
+	if (value_version) {
+		version = value_version.value();
+		if (version != 3) {
+			return MakeError(
+				VersionError,
+				"Failed to unmarshal the version from the version JSON data. Only 'version' 3 is supported received version "
+					+ std::to_string(version));
+		}
+	}
+
+	const json::ExpectedJson format_value = version_json.Get("format");
+	if (version_value) {
+		const json::ExpectedString value_format = format_value.value().GetString();
+		if (value_format) {
+			format = value_format.value();
+			if (format != "mender") {
+				return MakeError(
+					FormatError,
+					"Failed to unmarshal the 'format' from the version JSON data. The client only understands the 'mender' Artifact type. Got format: "
+						+ format);
+			}
+		}
+	}
+
+	return Version {.version = 3, .format = "mender"};
```

Comment:
It feels a bit strange to have a function called `Parse()` that doesn't actually return any parsed data, but not a big deal.

Reply:
Well, it does, right. The parsed values match exactly this. I just wanted to highlight it here. But if it's odd I can change it.